syntax = "proto3";
package edge;

message ImageRequest {
  string source = 1;
  string image_b64 = 2; // base64-encoded JPEG bytes
  int32 width = 3;
  int32 height = 4;
  int64 timestamp_ms = 5;
  int64 seq = 6; // client frame sequence id for RTT correlation
}

message Detection {
  int32 class_id = 1;
  float score = 2;
  float xmin = 3; // normalized 0..1
  float ymin = 4;
  float xmax = 5;
  float ymax = 6;
  string class_name = 7; // optional human-readable class name
}

message DetectionResponse {
  string source = 1;
  repeated Detection detections = 2;
  int64 timestamp_ms = 3;
  int64 seq = 4; // echoed client sequence id
  int64 processing_ms = 5; // server-side processing duration in ms
}

service Inference {
  // Bidirectional streaming: client streams images, server streams detections.
  rpc Stream(stream ImageRequest) returns (stream DetectionResponse);
}
